### PR TITLE
Add fund transaction creation

### PR DESCRIPTION
**Pull request by julia-zack**

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java [hunk 1]
```diff
@@ -748,6 +749,74 @@ private void logCommitmentWithVotedFederation(BridgeEventLogger eventLogger, Fed
...
+        int neededSignaturesSize = 72 * federation.getNumberOfSignaturesRequired();
+        return redeemScriptSize + neededSignaturesSize;
+    }
+
+    private SendRequest createSendRequest(BtcTransaction transaction, Coin feePerKb) {
+        SendRequest sendRequest = SendRequest.forTx(transaction);
+        sendRequest.changeAddress = getActiveFederationAddress();
+        sendRequest.feePerKb = feePerKb;
+        sendRequest.missingSigsMode = Wallet.MissingSigsMode.USE_OP_ZERO;
+        sendRequest.recipientsPayFees = false;
```
Comment (julia-zack) [inline rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java:815]: we dont want the proposed fed to pay for the fees since the amount to send was calculated quite precise

File: rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java [hunk 2]
```diff
@@ -748,6 +749,74 @@ private void logCommitmentWithVotedFederation(BridgeEventLogger eventLogger, Fed
...
+    }
+
+    private void addOutputToProposedFederation(BtcTransaction svpFundTransaction, Federation proposedFederation, Coin feePerKb) {
+        long transactionSizeToBeSentFromProposedFederation = calculateTransactionSizeToBeSentFromFederation(proposedFederation);
+        Coin neededValueToSendFromProposedFederation = feePerKb.multiply(transactionSizeToBeSentFromProposedFederation);
+
+        svpFundTransaction.addOutput(neededValueToSendFromProposedFederation, proposedFederation.getAddress());
+    }
+
+    private long calculateTransactionSizeToBeSentFromFederation(Federation federation) {
```
Comment (julia-zack) [inline rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java:782]: this calculation is a guide for now, it will be re-calculated in another task.

File: rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java (hunk 1 above)
Comment (marcos-iov) [reply rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java:778]: Interesting, this way no matter how high the current network fees are we ensure the proposed will receive enough funds to send back, right?

Let's take a note to double check this anyway

File: rskj-core/src/main/java/co/rsk/peg/bitcoin/BitcoinUtils.java [hunk 3]
```diff
@@ -61,4 +62,9 @@ public static Optional<Script> extractRedeemScriptFromInput(TransactionInput txI
             return Optional.empty();
         }
     }
+
+    public static Script createBaseInputScriptThatSpendsFromTheFederation(Federation federation) {
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/bitcoin/BitcoinUtils.java:66]: Is this still needed?

File: rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java [hunk 4]
```diff
@@ -747,6 +748,38 @@ private void logCommitmentWithVotedFederation(BridgeEventLogger eventLogger, Fed
         logger.debug("[logCommitmentWithVotedFederation] Voted federation committed: {}", votedFederation.getAddress());
     }
 
+    @Override
+    public BtcTransaction createAndSetSvpFundTransactionWithoutSignatures(Coin spendableValueFromProposedFederation, Coin feePerKb, Wallet activeFederationWallet) throws InsufficientMoneyException {
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java:752]: The fact that the method name has an `and` as part of it's name makes me think it's doing more than one thing and probably Uncle Bob does not approve. Should we split this into 2 separate methods?

File: rskj-core/src/main/java/co/rsk/peg/federation/FederationSupport.java [hunk 5]
```diff
@@ -53,5 +53,7 @@ public interface FederationSupport {
 
     void updateFederationCreationBlockHeights();
 
+    BtcTransaction createAndSetSvpFundTransactionWithoutSignatures(Coin spendableValueFromProposedFederation, Coin feePerKb, Wallet activeFederationWallet) throws InsufficientMoneyException;
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/federation/FederationSupport.java:56]: Maybe the `spendableValueFromProposedFederation` should come from BridgeConstants?

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 6]
```diff
@@ -962,6 +962,22 @@ public void updateCollections(Transaction rskTx) throws IOException {
         updateFederationCreationBlockHeights();
     }
 
+    protected void createAndProcessSvpFundTransactionWithoutSignatures() throws InsufficientMoneyException, IOException {
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:965]: Where is this used?

Same comment regarding the `and` in the method name, although in this case I think it'd be wiser to renamed rather than split.

Let's discuss this a bit, I guess it depends a bit on who the caller is going to be

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 7]
```diff
@@ -2191,6 +2195,50 @@ void clearRetiredFederation_whenHavingOldFederation_removesOldFederation() {
         assertThat(oldFederation, is(nullValue()));
     }
 
+    @Test
+    @Tag("create and process svp fund transaction unsigned")
+    void createAndProcessSvpFundTransactionWithoutSignatures_savesExpectedTransactionInStorage() throws InsufficientMoneyException {
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2200]: Any unexpected corner case we'd like to test?

No UTXOs available at the moment? Fees too high? InsufficientMoneyException?

File: rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java [hunk 8]
```diff
@@ -747,6 +748,38 @@ private void logCommitmentWithVotedFederation(BridgeEventLogger eventLogger, Fed
...
+            String message = "Proposed federation should be present when creating SVP fund transaction.";
+            logger.warn(message);
+            throw new IllegalStateException(message);
+        }
+        Federation proposedFederation = proposedFederationOpt.get();
+
+        BtcTransaction svpFundTransactionUnsigned = new BtcTransaction(constants.getBtcParams());
+        svpFundTransactionUnsigned.setVersion(BTC_TX_VERSION_2);
+
+        svpFundTransactionUnsigned.addOutput(spendableValueFromProposedFederation, proposedFederation.getAddress());
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java:764]: There's supposed to be an output for the flyover federation as well, right? Is that part of another task?

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 9]
```diff
@@ -2191,6 +2195,50 @@ void clearRetiredFederation_whenHavingOldFederation_removesOldFederation() {
...
+        // assert the amount sent was correct
+        TransactionOutput outputToProposedFederation = outputToProposedFederationOpt.get();
+        assertEquals(outputToProposedFederation.getValue(), spendableValueFromProposedFederation);
+
+        // assert all the other outputs are the change sent to the active federation
+        Script activeFederationScriptPubKey = activeFederation.getP2SHScript();
+        assertTrue(svpFundTransactionUnsignedOutputs.stream()
+            .filter(output -> !output.equals(outputToProposedFederation))
+            .allMatch(output -> output.getScriptPubKey().equals(activeFederationScriptPubKey)));
+    }
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2240]: assert the transaction is added to the pegout tx index?

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 6 above)
Comment (marcos-iov) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:965]: Should we emit release_requested event also? Where does that come into play? And `pegout_transaction_created` event

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 9 above)
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2240]: isn't that [this assertion](https://github.com/rsksmart/rskj/pull/2677/files#diff-bca5614b0a0ce7ba81d79322090b66cee8d9289e3430798d67767568abc8491aR637-R641)?

File: rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java (hunk 1 above)
Comment (julia-zack) [reply rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java:778]: i think that's always a risk, because the fees could vary between the fund transaction and the spend transaction are sent. But I wanted to make sure that the value chosen to be the one received from the proposed fed will actually be the value sent to it

File: rskj-core/src/main/java/co/rsk/peg/federation/FederationSupport.java (hunk 5 above)
Comment (julia-zack) [reply rskj-core/src/main/java/co/rsk/peg/federation/FederationSupport.java:56]: good one. Thanks.

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 6 above)
Comment (julia-zack) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:965]: the `protected` is for testing the method.
`release_requested` event can't be emitted because the rskTxHash is null right?
`pegout_transaction_created` should be added. Thanks!

Comment (julia-zack) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:965]: regarding the method name / behavior, yep maybe we can discuss it a bit :)

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 7 above)
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java]: mmm, I think that should be a test for `completeTx` method. Wdyt?

Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java]: i mean, if there are no utxos, `completeTx` should throw the exception. Is not being thrown from these new methods.

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 10]
```diff
@@ -962,6 +962,86 @@ public void updateCollections(Transaction rskTx) throws IOException {
...
+            String message = "The output for the proposed federation should be present in the svp fund transaction.";
+            logger.warn(message);
+            throw new IllegalStateException(message);
+        }
+        TransactionOutput outputToProposedFederation = outputToProposedFederationOpt.get();
+
+        return outputToProposedFederation.getValue();
+    }
+
+    private void performPegoutActions(BtcTransaction pegoutTransaction, Keccak256 rskTxHash, Coin requestedPegoutAmount) throws IOException {
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:1024]:
```suggestion
    private void confirmReleaseRequest(BtcTransaction pegoutTransaction, Keccak256 releaseCreationTxHash, Coin requestedAmount) throws IOException {
```
Tried to give a more meaningful name to the method and params. The idea is that this can be used for migration, pegouts, and fundTx right?

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 11]
```diff
@@ -962,6 +962,86 @@ public void updateCollections(Transaction rskTx) throws IOException {
...
+        addPegoutToPegoutsWaitingForConfirmations(pegoutTransaction, rskTxHash);
+        savePegoutTxSigHash(pegoutTransaction);
+        logReleaseRequested(rskTxHash.getBytes(), pegoutTransaction, requestedPegoutAmount);
+        logPegoutTransactionCreated(pegoutTransaction);
+    }
+
+    private void addPegoutToPegoutsWaitingForConfirmations(BtcTransaction pegoutTransaction, Keccak256 rskTxHash) throws IOException {
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:1031]: Isn't this gonna clash with the already existing method that has a very similar name?

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 12]
```diff
@@ -962,6 +962,86 @@ public void updateCollections(Transaction rskTx) throws IOException {
...
+        PegoutsWaitingForConfirmations pegoutsWaitingForConfirmations = provider.getPegoutsWaitingForConfirmations();
+        pegoutsWaitingForConfirmations.add(pegoutTransaction, rskExecutionBlock.getNumber(), rskTxHash);
+    }
+
+    private void logReleaseRequested(byte[] rskTxHashSerialized, BtcTransaction pegoutTransaction, Coin requestedPegoutAmount) {
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:1036]: Why not receive the hash object here and serialize it in the method? Seems like that's a responsibility of this method

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 13]
```diff
@@ -522,6 +553,223 @@ void save_callsFederationSupportSave() throws IOException {
...
+        private Sha256Hash svpFundTransactionHashUnsigned;
+
+        private Transaction rskTx;
+        private Keccak256 rskTxHash;
+
+        private PegoutsWaitingForConfirmations pegoutsWaitingForConfirmations;
+
+        @BeforeEach
+        void setUp() {
+            rskExecutionBlock = mock(Block.class);
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:573]: Use an actual block like we did in other tests?

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 14]
```diff
@@ -522,6 +553,223 @@ void save_callsFederationSupportSave() throws IOException {
...
+            long rskExecutionBlockNumber = 1000L;
+            when(rskExecutionBlock.getNumber()).thenReturn(rskExecutionBlockNumber);
+
+            activeFederation = FederationTestUtils.getGenesisFederation(federationConstantsMainnet);
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:577]: Not sure using the genesis federation is a good idea, it's a legacy standard federation. I'd use the current federation type

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 15]
```diff
@@ -522,6 +553,223 @@ void save_callsFederationSupportSave() throws IOException {
...
+            federationSupport = mock(FederationSupport.class);
+            when(federationSupport.getActiveFederation()).thenReturn(activeFederation);
+            when(federationSupport.getActiveFederationAddress()).thenReturn(activeFederation.getAddress());
+
+            List<UTXO> utxos = BitcoinTestUtils.createUTXOs(10, activeFederation.getAddress());
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:582]:
```suggestion
            List<UTXO> activeFederationUTXOs = BitcoinTestUtils.createUTXOs(10, activeFederation.getAddress());
```
Let's use meaningful variable names

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 16]
```diff
@@ -522,6 +553,223 @@ void save_callsFederationSupportSave() throws IOException {
...
+            when(federationSupport.getActiveFederationBtcUTXOs()).thenReturn(utxos);
+
+            proposedFederation = new P2shErpFederationBuilder().build();
+            when(federationSupport.getProposedFederation()).thenReturn(Optional.of(proposedFederation));
+
+            Coin feePerKb = Coin.valueOf(1000);
+            when(feePerKbSupport.getFeePerKb()).thenReturn(feePerKb);
+
+            rskTx = mock(Transaction.class);
+            rskTxHash = PegTestUtils.createHash3(1);
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:592]: Should this method be part of `RskTestUtils` actually?

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 17]
```diff
@@ -522,6 +553,223 @@ void save_callsFederationSupportSave() throws IOException {
...
+                .withProvider(bridgeStorageProvider)
+                .withActivations(activationsAfterForks)
+                .withFederationSupport(federationSupport)
+                .withFeePerKbSupport(feePerKbSupport)
+                .withExecutionBlock(rskExecutionBlock)
+                .build();
+        }
+
+        @Test
+        void createSvpFundTransactionWithoutSignatures_createsExpectedSvpFundTransaction() throws InsufficientMoneyException {
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:609]: Any corner cases to consider besides the happy path?

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 11 above)
Comment (julia-zack) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:1031]: yea i though about addressing that in the task in charge of doing that refactor. wdyt?

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 18]
```diff
@@ -962,6 +962,86 @@ public void updateCollections(Transaction rskTx) throws IOException {
...
+        Optional<Federation> proposedFederationOpt = federationSupport.getProposedFederation();
+        if (!proposedFederationOpt.isPresent()) {
+            String message = "Proposed federation should be present when processing SVP fund transaction.";
+            logger.warn(message);
+            throw new IllegalStateException(message);
+        }
+        Federation proposedFederation = proposedFederationOpt.get();
+
+        BtcTransaction svpFundTransactionUnsigned = createSvpFundTransactionWithoutSignatures(proposedFederation);
+        provider.setSvpFundTxHashUnsigned(svpFundTransactionUnsigned.getHash());
```
[the remark this answers is not in the document]
Comment (julia-zack) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:975]: probably the same, but lets analyze that better in a new task
